lldb/include/lldb/Target/Statistics.h

```diff
@@ -174,6 +177,83 @@ struct StatisticsOptions {
   std::optional<bool> m_include_transcript;
 };
 
+/// A class that represents statistics about a TypeSummaryProviders invocations
+class SummaryStatistics {
+public:
+  explicit SummaryStatistics(std::string name, std::string impl_type)
+      : m_total_time(), m_impl_type(impl_type), m_name(name),
+        m_count(0) {}
+
+  std::string GetName() const { return m_name; };
+  double GetTotalTime() const { return m_total_time.get().count(); }
+
+  uint64_t GetSummaryCount() const {
+    return m_count.load(std::memory_order_relaxed);
+  }
+
+  StatsDuration &GetDurationReference() { return m_total_time; };
+
+  std::string GetSummaryKindName() const { return m_impl_type; }
+
+  llvm::json::Value ToJSON() const;
+
+
+  /// Basic RAII class to increment the summary count when the call is complete.
+  /// In the future this can be extended to collect information about the
+  /// elapsed time for a single request.
+  class SummaryInvocation{
+  public:
+    SummaryInvocation(std::shared_ptr<SummaryStatistics> summary_stats)
+        : m_stats(summary_stats),
+          m_elapsed_time(summary_stats->GetDurationReference()) {}
+    ~SummaryInvocation() { m_stats->OnInvoked(); }
+
+    /// Delete the copy constructor and assignment operator to prevent
+    /// accidental double counting.
+    /// @{
+    SummaryInvocation(const SummaryInvocation &) = delete;
+    SummaryInvocation &operator=(const SummaryInvocation &) = delete;
+    /// @}
+
+  private:
+    std::shared_ptr<SummaryStatistics> m_stats;
+    ElapsedTime m_elapsed_time;
+  };
+
+private:
+  void OnInvoked() {
+    m_count.fetch_add(1, std::memory_order_relaxed);
+  }
+  lldb_private::StatsDuration m_total_time;
+  std::string m_impl_type;
+  std::string m_name;
+  std::atomic<uint64_t> m_count;
+};
+
+typedef std::shared_ptr<SummaryStatistics> SummaryStatisticsSP;
+
+/// A class that wraps a std::map of SummaryStatistics objects behind a mutex.
+class SummaryStatisticsCache {
+public:
+
+  /// Get the SummaryStatistics object for a given provider name, or insert
+  /// if statistics for that provider is not in the map.
+  SummaryStatisticsSP
+  GetSummaryStatisticsForProviderName(lldb_private::TypeSummaryImpl &provider) {
+    std::lock_guard<std::mutex> guard(m_map_mutex);
+    auto pair = m_summary_stats_map.try_emplace(
```

Comment:
@mbucko / @clayborg 

I wanted to still use try emplace here, but I would need the RHS to be a `shared_ptr<T> &&`, which will require a construction of an SP to even try and insert. That call to new seems to defeat the purpose of avoiding a 'lookup and insert if not present' optimization. Can I get some advice on a better impl here?

Reply:
We don't want to construct a new shared pointer each time with try_emplace, so just do a `llvm::StringMap::find()` first, and if is succeeds it will return a valid shared pointer and you can return that. If it doesn't find it, you can use the existing code, and possibly just use `emplace` right?

```
SummaryStatisticsSP
  GetSummaryStatisticsForProviderName(lldb_private::TypeSummaryImpl &provider) {
    std::lock_guard<std::mutex> guard(m_map_mutex);
    if (SummaryStatisticsSP summary_stats_sp = m_summary_stats_map.find(provider.GetName()))
      return summary_stats_sp;
    auto pair = m_summary_stats_map.try_emplace(
      provider.GetName(), 
      std::make_shared<SummaryStatistics>(provider.GetName(), 
                                                                          provider.GetSummaryKindName()));
    return pair.first->second;
}
```

Reply:
Remember that `llvm::StringMap::find()` will return a default constructed shared pointer, which will be empty.